protocols/raft/src/main/java/io/atomix/protocols/raft/service/impl/DefaultServiceContext.java

```diff
@@ -317,23 +317,33 @@ private void maybeInstallSnapshot(long index) {
       // Compute the snapshot index as the greater of the compaction index and the last index applied to this service.
       long snapshotIndex = Math.max(index, currentIndex);
 
+      // If a snapshot is already persisted at this index, skip the snapshot.
+      Snapshot existingSnapshot = raft.getSnapshotStore().getSnapshot(serviceId, snapshotIndex);
+      if (existingSnapshot != null) {
+        return;
```

Comment:
Shouldn't we be completing the future when returning early?

Reply:
Yep...

Reply:
Fixed